Refactor RBAC authorizer entry points

enj opened the pull request:
This change refactors various RBAC authorizer functions to be more flexible in their inputs.  This makes it easier to reuse the various components that make up the authorizer.

Signed-off-by: Monis Khan <mkhan@redhat.com>

```release-note
NONE
```

enj commented:
@kubernetes/sig-auth-pr-reviews 
/sig auth
/assign @deads2k

deads2k commented on plugin/pkg/auth/authorizer/rbac/rbac.go:
```diff
@@ -121,13 +120,8 @@ func (r *RBACAuthorizer) Authorize(requestAttributes authorizer.Attributes) (boo
-		),
-	}
-	return authorizer
+func New(ruleResolver RequestToRuleMapper) *RBACAuthorizer {
```
I don't like coercing the interfaces this way.  I do appreciate the difficulty in casting the listers (which you have) to the rbac interfaces (which you don't have), but I'd rather see the shims moved and be exposed here.

enj committed:
Refactor RBAC authorizer entry points

This change refactors various RBAC authorizer functions to be more
flexible in their inputs.  This makes it easier to reuse the various
components that make up the authorizer.

Signed-off-by: Monis Khan <mkhan@redhat.com>

enj commented:
/retest

enj commented:
/retest

deads2k commented on plugin/pkg/auth/authorizer/rbac/subject_locator.go:
```diff
@@ -39,16 +45,13 @@ type SubjectAccessEvaluator struct {
 
-func NewSubjectAccessEvaluator(roles rbacregistryvalidation.RoleGetter, roleBindings rbacregistryvalidation.RoleBindingLister, clusterRoles rbacregistryvalidation.ClusterRoleGetter, clusterRoleBindings rbacregistryvalidation.ClusterRoleBindingLister, superUser string) *SubjectAccessEvaluator {
-	subjectLocator := &SubjectAccessEvaluator{
+func NewSubjectAccessEvaluator(roleBindings rbacregistryvalidation.RoleBindingLister, clusterRoleBindings rbacregistryvalidation.ClusterRoleBindingLister, roleToRuleMapper RoleToRuleMapper, superUser string) *SubjectAccessEvaluator {
```
I'd keep this signature as it was.

deads2k commented:
/lgtm

deads2k commented:
/approve no-issue
